### PR TITLE
[Fixtures] Customer factory

### PR DESCRIPTION
**Pull request by loic425**
| Q               | A
| --------------- | -----
| Branch?         | foundry-fixtures
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | 
| License         | MIT

### CODE REVIEW

File: psalm.xml [hunk 1]
```diff
@@ -148,6 +148,12 @@
             </errorLevel>
         </UndefinedDocblockClass>
 
+        <UndefinedMethod>
+            <errorLevel type="info">
+                <file name="src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/ShopUserFactory.php" />
```
Comment (vvasiloi) [inline psalm.xml:153]: What is the method? I would rather add the annotations inline.

File: src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/ShopUserFactory.php [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\CoreBundle\DataFixtures\Factory;
+
+use Sylius\Component\Core\Model\ShopUser;
+use Sylius\Component\Core\Model\ShopUserInterface;
+use Sylius\Component\Customer\Model\CustomerGroupInterface;
+use Sylius\Component\Customer\Model\CustomerInterface;
```
Comment (vvasiloi) [inline src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/ShopUserFactory.php:19]:
```suggestion
use Sylius\Component\Core\Model\CustomerInterface;
```

File: src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/ShopUserFactory.php [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+            return ['birthday' => $birthday];
+        });
+    }
+
+    public function withPassword(string $password): self
+    {
+        return $this->addState(['password' => $password]);
+    }
+
+    public function withCustomerGroup(Proxy|CustomerGroupInterface|string $customerGroup): self
```
Comment (vvasiloi) [inline src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/ShopUserFactory.php:97]: Is the `Proxy` type-hint really necessary? The proxy should implement `CustomerGroupInterface` and should pass the type check.

File: src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/ShopUserFactory.php [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        return [
+            'email' => self::faker()->email(),
+            'first_name' => self::faker()->firstName(),
+            'last_name' => self::faker()->lastName(),
+            'enabled' => true,
+            'password' => 'password123',
+            'customer_group' => $this->customerGroupFactory::randomOrCreate(),
+            'gender' => CustomerInterface::UNKNOWN_GENDER,
+            'phone_number' => self::faker()->phoneNumber(),
+            'birthday' => self::faker()->dateTimeThisCentury(),
```
Comment (vvasiloi) [inline src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/ShopUserFactory.php:119]: Very young audience. How about generating them of ages between 18 and 80 years: `->dateTimeBetween('-80 years', '-18 years')`?

File: src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/ShopUserFactoryInterface.php [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+{
+    public function withEmail(string $email): self;
+
+    public function withFirstName(string $firstName): self;
+
+    public function withLastName(string $lastName): self;
+
+    public function withCustomerGroup(Proxy|CustomerGroupInterface|string $customerGroup): self;
+
+    public function male(): self;
```
Comment (vvasiloi) [inline src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/ShopUserFactoryInterface.php:48]: missing female method

File: psalm.xml (hunk 1 above)
Comment (loic425) [reply psalm.xml:153]: `findOrCreate` method
https://github.com/Sylius/Sylius/pull/13543/files#diff-db5036129c448c53aa427624ed5e44a5176ac04fa0d4849a1bf183629a618534R30

File: src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/ShopUserFactory.php (hunk 3 above)
Comment (loic425) [reply src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/ShopUserFactory.php:97]: Argument #1 ($customerGroup) must be of type Sylius\Component\Customer\Model\CustomerGroupInterface|string, Zenstruck\Foundry\Proxy given
So it does not implement the interface.

File: src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/ShopUserFactory.php (hunk 4 above)
Comment (loic425) [reply src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/ShopUserFactory.php:119]: cause it's the current implementation, but you're right, it makes sense. @lchrusciel @Zales0123 What do you think?

https://github.com/Sylius/Sylius/blob/master/src/Sylius/Bundle/CoreBundle/Fixture/Factory/ShopUserExampleFactory.php#L87

File: src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/ShopUserFactoryInterface.php (hunk 5 above)
Comment (loic425) [reply src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/ShopUserFactoryInterface.php:48]: I'm always impressed for your review quality. That's great.

File: src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/ShopUserFactory.php [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    }
+
+    public function withPhoneNumber(string $phoneNumber): self
+    {
+        return $this->addState(['phone_number' => $phoneNumber]);
+    }
+
+    public function withBirthday(\DateTimeInterface|string $birthday): self
+    {
+        return $this->addState(function () use ($birthday): array {
```
[the remark this answers is not in the document]
Comment (vvasiloi) [reply src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/ShopUserFactory.php:83]: I think it is to generate a new _default_ value every time. It seems to be similar to how Sylius sets default in current fixture factories:
https://github.com/Sylius/Sylius/blob/56bf4199f4c63e0c9101b1e3eb42bd5e4553583a/src/Sylius/Bundle/CoreBundle/Fixture/Factory/ShopUserExampleFactory.php#L72-L74

If so, then it doesn't make much sense to use it for methods that only operate with explicit values.

File: src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/ShopUserFactoryInterface.php (hunk 5 above)
Comment (loic425) [reply src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/ShopUserFactoryInterface.php:48]: done

File: src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/ShopUserFactory.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (loic425) [reply src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/ShopUserFactory.php]: Thank you for the explanation, that makes sense.

I changed that.

File: psalm.xml (hunk 1 above)
Comment (loic425) [reply psalm.xml:153]: I don't know why Psalm does not look at the docblock

File: src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/ShopUserFactoryInterface.php (hunk 5 above)
Comment (vvasiloi) [reply src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/ShopUserFactoryInterface.php:48]: Thank you! 😅

File: psalm.xml (hunk 1 above)
Comment (vvasiloi) [reply psalm.xml:153]: Looks like it only works if the `__call` method is implemented: https://psalm.dev/docs/annotating_code/supported_annotations/#supported-docblock-annotations

File: src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/ShopUserFactory.php (hunk 4 above)
Comment (Zales0123) [reply src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/ShopUserFactory.php:117]: 👍 for the change, especially in the new implementation. In the current one, you can open a PR and we will see ;)

Comment (loic425) [reply src/Sylius/Bundle/CoreBundle/DataFixtures/Factory/ShopUserFactory.php]: @vvasiloi done

**Pull request merged**